File: leet/0230-kth-smallest-element-in-a-BST/kth-smallest-element-in-a-BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        
        p = root
        stack = []
        
        while p is not None:
            stack.append(p)
            p = p.left
            
        while len(stack) > 0:
            p = stack.pop()
            k -= 1
            
            if k == 0 :
                return p.val
            
            p = p.right
            
            while p is not None:
                stack.append(p)
                p = p.left

File: leet/0230-kth-smallest-element-in-a-BST/kth-smallest-element-in-a-BST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function(root, k) {
    let stack = []
    let cur = root
    while(cur){
        stack.push(cur)
        cur = cur.left
    }
    
    while(stack.length > 0){
        cur = stack.pop()        
        if(k == 1){
            return cur.val
        }
        
        k -= 1
        if(cur.right){
            cur = cur.right
            while(cur){
                stack.push(cur)
                cur = cur.left
            }
        }
    }
    
    return -1
};
